File: backend/schemas/analytics.py
from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel, Field


class AnalyticsTotals(BaseModel):
  flow_count: int = 0
  total_bytes: int = 0
  total_packets: int = 0
  threat_count: int = 0
  blocked_count: int = 0
  allowed_count: int = 0
  unique_source_ips: int = 0


class AnalyticsTimePoint(BaseModel):
  period: str
  flow_count: int = 0
  total_bytes: int = 0
  total_packets: int = 0
  threat_count: int = 0
  blocked_count: int = 0
  allowed_count: int = 0
  unique_source_ips: int = 0


class DailyAnalyticsResponse(BaseModel):
  success: bool = True
  message: str
  days: int
  start_date: date
  end_date: date
  data_points: list[AnalyticsTimePoint]
  totals: AnalyticsTotals
  generated_at: datetime


class MonthlyAnalyticsResponse(BaseModel):
  success: bool = True
  message: str
  months: int
  start_period: str
  end_period: str
  data_points: list[AnalyticsTimePoint]
  totals: AnalyticsTotals
  generated_at: datetime


class ProtocolAnalyticsItem(BaseModel):
  protocol: str
  flow_count: int
  total_bytes: int
  total_packets: int
  threat_count: int
  blocked_count: int
  percentage: float


class ProtocolAnalyticsResponse(BaseModel):
  success: bool = True
  message: str
  period_hours: int
  total_flows: int
  total_bytes: int
  protocols: list[ProtocolAnalyticsItem]
  generated_at: datetime


class CountryAnalyticsItem(BaseModel):
  country_code: str
  country_name: str
  flow_count: int
  total_bytes: int
  threat_count: int
  blocked_count: int
  percentage: float
  top_source_ips: list[str] = Field(default_factory=list)


class CountryAnalyticsResponse(BaseModel):
  success: bool = True
  message: str
  period_hours: int
  total_flows: int
  total_bytes: int
  countries: list[CountryAnalyticsItem]
  unknown_flow_count: int = 0
  generated_at: datetime


class AnalyticsOverviewResponse(BaseModel):
  success: bool = True
  message: str
  daily: DailyAnalyticsResponse
  monthly: MonthlyAnalyticsResponse
  protocols: ProtocolAnalyticsResponse
  countries: CountryAnalyticsResponse
  generated_at: datetime

File: frontend/src/types/analytics.ts
export interface AnalyticsTotals {
  flow_count: number;
  total_bytes: number;
  total_packets: number;
  threat_count: number;
  blocked_count: number;
  allowed_count: number;
  unique_source_ips: number;
}

export interface AnalyticsTimePoint {
  period: string;
  flow_count: number;
  total_bytes: number;
  total_packets: number;
  threat_count: number;
  blocked_count: number;
  allowed_count: number;
  unique_source_ips: number;
}

export interface DailyAnalytics {
  success: boolean;
  message: string;
  days: number;
  start_date: string;
  end_date: string;
  data_points: AnalyticsTimePoint[];
  totals: AnalyticsTotals;
  generated_at: string;
}

export interface MonthlyAnalytics {
  success: boolean;
  message: string;
  months: number;
  start_period: string;
  end_period: string;
  data_points: AnalyticsTimePoint[];
  totals: AnalyticsTotals;
  generated_at: string;
}

export interface ProtocolAnalyticsItem {
  protocol: string;
  flow_count: number;
  total_bytes: number;
  total_packets: number;
  threat_count: number;
  blocked_count: number;
  percentage: number;
}

export interface ProtocolAnalytics {
  success: boolean;
  message: string;
  period_hours: number;
  total_flows: number;
  total_bytes: number;
  protocols: ProtocolAnalyticsItem[];
  generated_at: string;
}

export interface CountryAnalyticsItem {
  country_code: string;
  country_name: string;
  flow_count: number;
  total_bytes: number;
  threat_count: number;
  blocked_count: number;
  percentage: number;
  top_source_ips: string[];
}

export interface CountryAnalytics {
  success: boolean;
  message: string;
  period_hours: number;
  total_flows: number;
  total_bytes: number;
  countries: CountryAnalyticsItem[];
  unknown_flow_count: number;
  generated_at: string;
}

export interface AnalyticsOverview {
  success: boolean;
  message: string;
  daily: DailyAnalytics;
  monthly: MonthlyAnalytics;
  protocols: ProtocolAnalytics;
  countries: CountryAnalytics;
  generated_at: string;
}

export interface AnalyticsFiltersState {
  days: number;
  months: number;
  periodHours: number;
}

export const DEFAULT_ANALYTICS_FILTERS: AnalyticsFiltersState = {
  days: 30,
  months: 12,
  periodHours: 24,
};

export type AnalyticsDashboardTab = "charts" | "tables";
